### ISSUE TITLE
Could view linting be suppressing contributions?

### ISSUE THREAD

**Issue by benhalpern** [thepracticaldev/dev.to#1828]
I have the feeling that view linting could be presenting a pretty strong burden to contributions. You could make a small contribution and get hung up with a huge linting fix chore as it currently stands.

I just feel like as an outsider without context it could add a bit too much burden at the time being. Thoughts on taking it away as a blocker to pushing and leave it as something folks can voluntarily make progress on until the app is at a closer place to being _fully linted_ on the view with our preferred styles?

**Comment by rhymes** [comment]
You make a really good point!

I don't know if it's actually deterring people from contributing but I agree that it can be a nuisance when ran automatically, especially if the contributor is a beginner.

I wonder if, before taking a decision either way, we could run the linting on the view files and see how far along the whole thing is and/or if Code climate has a specific metric on those.

This way it can be decided that until it reaches, for example 90+% of linting, it gets disabled as a git hook, to be re-enabled afterwards... ?

The downside of disabling view linting altogether is that people need to open a PR to fix linting, either with small mergeable increments or a quick one off. Because linting only PRs kept open too long tend to conflict with feature changes.

**Comment by lightalloy** [comment]
Ben, you mean that someone makes a small contribution and has to fix all the issues in the changed file?
That could be a problem, but I'm not sure about it.

In my opinion, the code will never reach 90+-% (or similar) of linting w/o the hook that forces the "correct" code style, because the new code won't be checked also, and the contributors (even the core team) won't check it every time unless the check is forced. So the gradual fixing from time to time won't be effective probably.

One of the solutions could be fixing all issues at once, but this solution has its own disadvantages: spending significant time on this and possible conflicts that would need to be fixed.

**Comment by Glennmen** [comment]
I think Ben is referring to what I mentioned here #1784 and that was only from one view file.

Also I wrote a little about it in [my blog post](https://dev.to/glennmen/changelog-github-pr-liquid-tag-support-3m1b) and tried to open a discussion about it, glad to see that Ben is doing that here.

**Comment by lightalloy** [comment]
@Glennmen Thanks for starting a discussion.
As for the existing linting issues in views, there are a lot of them, cause views linting has been introduced recently. The hook should only check the new or changed files before the commit. For now, I'm not sure why it required you to fix an untouched view file.
The `--no-verify` is probably used sometimes. In my opinion, it's not ok in most cases. I've noticed that several new `rubocop` issues appeared since I have checked (and fixed) the last time. In my opinion, it's better to add exceptions to the `rubobop_todo` or use `rubocop:disable/enable` at least, than run with `--no-verify`.

**Comment by Glennmen** [comment]
@lightalloy Just to be clear, I did edit that view file. I added the Github PR info to the markdown help docs.
However I still had to add `--no-verify` or else I wasn't able to commit, there was 1 linting error that I wasn't able to fix because it would parse the html in the code block instead of print it and I wasn't able to find a fix for that.

**Comment by lightalloy** [comment]
@Glennmen, oh, I see. Thanks for the explanation. I've made [a pr](https://github.com/thepracticaldev/dev.to/pull/1830) to disable the corresponding cops.

**Comment by nickytonline** [comment]
I have no idea if it's preventing contributions. Could be a good meta post @benhalpern on the site to ask the community.

Having said that, I think the issue is not the pre-commit, but the fact that there is linting debt and running `--no-verify` adds to that. If someone were to create new files, so no previous lint debt, and linting fails, I consider that a good thing. The linting rules (eslint, rubocop) are in place on the front-end and back-end to help make the code more maintainable.

The other thing is the pre-commit tries to fix linting errors, hence the `--fix`, so if there are still other errors, it's really because linting rules were violated.

I think enforcing these rules, assuming we fix existing linting debt is a good thing. It's not uncommon in the real world (open source or closed source) to fail at the pre-commit or PR build stage because of linting.

I would approach it this way:
- Look at the existing rules. Are we happy with them? Do we want to drop some?
- Assess the lint debt, i.e. which files fail linting currently.
- Fix them (could be a big job potentially), but an offshoot of this is these are actually potentially good first issues.

Just tor reiterate, this codebase will probably live a long life, so whatever we can do to keep it maintainable in an automated way is a good thing.

![](https://media.giphy.com/media/CmFMWpEa4IFtS/giphy-downsized.gif)

**Comment by benhalpern** [comment]
@nickytonline I made a post to ask the question in general. It could add to the discussion here. https://dev.to/ben/can-forced-linting-surpress-contributions-when-linting-is-first-introduced-2dn5

I agree with the longterm benefits of linting, of course, but I feel like we _may_ want to introduce this particular hook in some kind of more gradual way. Perhaps as a voluntary option at first.

Anyway, I’m not super strongly opinionated on this, I just wanted to get the concern out of my head. I’ll let @lightalloy and @maestromac make whatever choice they feel is appropriate. No rush.

**Comment by Glennmen** [comment]
I already planned to make an issue about something else I found before this issue but since it is kinda related I will add it to the discussion here (if wanted I could still make a separate issue).
This was something that wasn't being catched by the pre-commit linting, travis, codeclimate, ... but I only noticed it because of my IDE (RubyMine/JetBrains). None of the users probably noticed because luckily browsers are very "forgiving' and are able to parse the HTML even with errors. 

Besides the linting errors that are being thrown now there are tons of HTML errors in loads of view files, some examples:
Different opening and closing tags:
https://github.com/thepracticaldev/dev.to/blob/e676d9fc42cb6acdb3476abc4b7074cda676edaf/app/views/pages/_publishing_from_rss_guide_text.html.erb#L6

No `<dl>` closing tag (multiple times in the same file):
https://github.com/thepracticaldev/dev.to/blob/e676d9fc42cb6acdb3476abc4b7074cda676edaf/app/views/pages/_editor_guide_text.html.erb#L173-L178

No `<h4>` closing tag:
https://github.com/thepracticaldev/dev.to/blob/e676d9fc42cb6acdb3476abc4b7074cda676edaf/app/views/pages/generator.html.erb#L42

I think you get the picture 😉 I would like to go through the files in `/app/view/*` and fix any HTML errors (and linting errors while I'm at it). But need to know a couple of things first:
- How to submit?
  - 1 Big PR
  - Separate PR's for each folder (combined by first subfolder, nested folders included in that PR)
- Linting errors that I can't fix?
  - Ignore them
  - Notify someone from DEV? to change lint rules and merge master to my branch to commit without errors.

Currently these are the only questions that I can think of.

So what you guys think?

**Comment by aspittel** [comment]
I would lean towards disabling this until refactoring/cleaning up of older code is complete. Having a bunch of errors intimidated me when I started working on the code base, and I think it might for other people too. Refactoring some of the Preact code to fit the linting requirements would be pretty time consuming with the amount of errors and the depth of them.

**Comment by maestromac** [comment]
I like @nickytonline 's approach a lot. 

@Glennmen 1 big PR for fixing lint is fine but please split it up if it's more than 1000 line changes. Please ignore what you can't fix but also let us know what those are so we can decide if something should be changed.

**Comment by Zhao-Andy** [comment]
Also in favor of disabling until a refactoring/linting is done, like @aspittel said. Following @nickytonline's approach seems to be the right direction too.

It might be worthwhile to look at this from an 80/20 perspective, where we spend some time to resolve linting errors for the easiest 80% of the files (all `.scss` and `.html.erb` files, for example), and leave the hardest 20% to be refactored properly (Webpack/Preact/`.jsx` files, for example) another time. `.js.erb` (Javascript in asset pipeline), `.js` and `.jsx` (Webpack) files will require more time and consideration to refactor properly. 

I believe @lightalloy and @rhymes (and probably others I'm forgetting, sorry!) refactored a bunch of cops for Ruby files already, and we have a nice `.rubocop_todo.yml` file to work off of now. I think resolving the HTML and SCSS files will be pretty straightforward.

**Comment by rhymes** [comment]
I've tried running `erblint --lint-all --autocorrect` on the entire code base and I got the following output:

```shell
$ time erblint --lint-all --autocorrect
Linting 278 files with 11 autocorrectable linters...
[...]

# HERE THERE IS A LIST OF FILES THAT CAN'T BE AUTOCORRECTED

2010 error(s) corrected in ERB files
107.93s user 4.71s system 119.33s real 84308kB mem
```

The `git diff` is quite big and completely generated by the autocorrect. Mostly spaces, fixed indentation,  hash parameters and other small things but I also see some questionable refactorings.

This is how many lines have been changed:

```shell
$ git diff --shortstat
 199 files changed, 1460 insertions(+), 1521 deletions(-)
```

I **would not** recommend doing a big PR with the autocorrect, but maybe limiting the scope to a small amount of files for each PR should be doable. So that one can manually test if there's anything broken in the pages.

I've attached the output of `git diff --patch` if you want to have a look

[erblint.patch.txt](https://github.com/thepracticaldev/dev.to/files/2881218/erblint.patch.txt)

**Comment by Glennmen** [comment]
From rhymes his intel I would say it might be to big for 1 PR, especially since those numbers are only lint fixes and doesn't include any HTML fixes that I also want to include in the PR.

**Comment by lightalloy** [comment]
Please, be careful with `--autocorrect`. It's not ideal for `erblint`, I found out that sometimes it tries to fix the indentation incorrectly, which leads to invalid code. So the autocorrected files should be checked before committing.
I agree with @rhymes that it would be better to make relatively small prs to eliminate the debt gradually,

**Comment by Glennmen** [comment]
I will probably start fixing the view files this weekend, unless someone decides it needs more thought first.
Because of the size of the changes this is my planned workflow:
- Each folder in `/app/views/` will be a separate PR, so 32 PR's will be made. Some might be small PR's with only a couple of changes but splitting this up will help us review quicker and get it all merged.
- I will only do `html.erb` files, so no javascript or json files if there are any in the views folder.
- I will not do `--autocorrect`, I am going to go through each file manually and fix any linting and html errors.
- If there any linting errors that I can't fix, I will mention it in the PR draft/WIP and only after all errors are fixed I will make it a release ready PR.

Any feedback from you guys is welcome, especially to keep it easy for you guys to review it.

**Comment by lightalloy** [comment]
@Glennmen Using `--autocorrect` is fine, if you check the fixes manually afterward. It fixes most of the cases correctly and could make the process much faster.

**Comment by lightalloy** [comment]
@Glennmen I've also made an issue for the task.
